chain/db/store.go

```diff
@@ -280,46 +277,80 @@ func (cs *ChainStore) persist(b *Block) error {
 				amountSender := accSender.GetBalance()
 				donation, err := cs.GetDonation()
 				if err != nil {
 					return err
 				}
 				accSender.SetBalance(amountSender - donation.Amount)
-				states.SetAccount(BytesToUint160(coinbase.Sender), accSender)
 			}
+			nonce := accSender.GetNonce()
+			accSender.SetNonce(nonce + 1)
+			states.SetAccount(BytesToUint160(coinbase.Sender), accSender)
 
 			acc := states.GetOrNewAccount(BytesToUint160(coinbase.Recipient))
 			amount := acc.GetBalance()
-			acc.SetBalance(amount + Fixed64(coinbase.Amount))
+			acc.SetBalance(amount + Fixed64(coinbase.Amount) + totalFee)
 			states.setAccount(BytesToUint160(coinbase.Recipient), acc)
 		case TransferAssetType:
 			transfer := pl.(*TransferAsset)
 			accSender := states.GetOrNewAccount(BytesToUint160(transfer.Sender))
 			amountSender := accSender.GetBalance()
-			accSender.SetBalance(amountSender - Fixed64(transfer.Amount))
-			states.setAccount(BytesToUint160(transfer.Sender), accSender)
+			accSender.SetBalance(amountSender - Fixed64(transfer.Amount) - Fixed64(txn.UnsignedTx.Fee))
+			nonce := accSender.GetNonce()
+			accSender.SetNonce(nonce + 1)
+			states.SetAccount(BytesToUint160(transfer.Sender), accSender)
 
 			accRecipient := states.GetOrNewAccount(BytesToUint160(transfer.Recipient))
 			amountRecipient := accRecipient.GetBalance()
 			accRecipient.SetBalance(amountRecipient + Fixed64(transfer.Amount))
 			states.setAccount(BytesToUint160(transfer.Recipient), accRecipient)
 		case RegisterNameType:
 			registerNamePayload := pl.(*RegisterName)
+			pg, err := txn.GetProgramHashes()
+			if err != nil {
+				return err
+			}
+			accSender := states.GetOrNewAccount(pg[0])
+			amountSender := accSender.GetBalance()
+			accSender.SetBalance(amountSender - Fixed64(txn.UnsignedTx.Fee))
+			nonce := accSender.GetNonce()
+			accSender.SetNonce(nonce + 1)
+			states.SetAccount(pg[0], accSender)
 			err = cs.SaveName(registerNamePayload.Registrant, registerNamePayload.Name)
 			if err != nil {
 				return err
 			}
 		case DeleteNameType:
 			deleteNamePayload := pl.(*DeleteName)
+			pg, err := txn.GetProgramHashes()
+			if err != nil {
+				return err
+			}
+			accSender := states.GetOrNewAccount(pg[0])
+			amountSender := accSender.GetBalance()
+			accSender.SetBalance(amountSender - Fixed64(txn.UnsignedTx.Fee))
+			nonce := accSender.GetNonce()
+			accSender.SetNonce(nonce + 1)
+			states.SetAccount(pg[0], accSender)
 			err = cs.DeleteName(deleteNamePayload.Registrant)
 			if err != nil {
 				return err
 			}
 		case SubscribeType:
 			subscribePayload := pl.(*Subscribe)
+			pg, err := txn.GetProgramHashes()
+			if err != nil {
+				return err
+			}
+			accSender := states.GetOrNewAccount(pg[0])
+			amountSender := accSender.GetBalance()
+			accSender.SetBalance(amountSender - Fixed64(txn.UnsignedTx.Fee))
+			nonce := accSender.GetNonce()
+			accSender.SetNonce(nonce + 1)
+			states.SetAccount(pg[0], accSender)
```

Comment:
There are quite a lot duplicated codes here. What about creating a function to call these common codes and just call the function here?

Reply:
em. I do agree with you.